File: 309. Best Time to Buy and Sell Stock with Cooldown/C++/Solution.cpp
class Solution {
public:
    int maxProfit(vector<int> &prices) {
        if (prices.empty()) {
            return 0;
        }
        vector<int> buy(prices.size());
        vector<int> sell(prices.size());
        buy[0] = -prices[0];
        sell[0] = 0;
        for (int i = 1; i < prices.size(); i++) {
            if (i == 1) {
                buy[i] = max(buy[i - 1], -prices[i]);
            } else {
                buy[i] = max(buy[i - 1], sell[i - 2] - prices[i]);
            }
            sell[i] = max(sell[i - 1], buy[i - 1] + prices[i]);
        }
        return sell[prices.size() - 1];
    }
};

File: 309. Best Time to Buy and Sell Stock with Cooldown/Python/Solution.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if not prices:
            return 0
        buy = [0] * len(prices)
        sell = [0] * len(prices)
        buy[0] = -prices[0]
        for i in range(1, len(prices)):
            if i == 1:
                buy[i] = max(buy[i - 1], -prices[i])
            else:
                buy[i] = max(buy[i - 1], sell[i - 2] - prices[i])
            sell[i] = max(sell[i - 1], buy[i - 1] + prices[i])
        return sell[len(prices) - 1]
